src/main/java/org/deepsymmetry/beatlink/DeviceFinder.java

```diff
@@ -180,268 +134,232 @@ public boolean isAddressIgnored(InetAddress address) {
-     * reachable.
+     * Start listening for device announcements and keeping track of the DJ Link devices visible on the network.
+     * If already listening, has no effect.
+     * @param  announcement handles announcement statuses which have already expired and added new devices.
      */
-    synchronized void flush() {
-        final Set<DeviceAnnouncement> lastDevices = new HashSet<DeviceAnnouncement>(devices.values());
-        devices.clear();
-        firstDeviceTime.set(0);
+    public synchronized void handleDeviceStatusAnnouncement(DeviceAnnouncement announcement) {
+        if (isRunning()) {
 
-        // Report the loss of all our devices, on the proper thread, also outside our lock.
-        SwingUtilities.invokeLater(new Runnable() {
-            @Override
-            public void run() {
-                for (DeviceAnnouncement announcement : lastDevices) {
-                    deliverLostAnnouncement(announcement);
-                }
+            final boolean foundNewDevice = isDeviceNew(announcement);
+            updateDevices(announcement);
+            if (foundNewDevice) {
+                deliverFoundAnnouncement(announcement);
+            }
+            if (VirtualCdj.getInstance().isRunning() &&
+                    announcement.getDeviceNumber() == VirtualCdj.getInstance().getDeviceNumber()) {
+                // Someone is using the same device number as we are! Try to defend it.
+                VirtualCdj.getInstance().defendDeviceNumber(announcement.getAddress());
             }
-        });
-    }
-
-    /**
-     * Stop listening for device announcements. Also discard any announcements which had been received, and
-     * notify any registered listeners that those devices have been lost.
-     */
-    @SuppressWarnings("WeakerAccess")
-    public synchronized void stop() {
-        if (isRunning()) {
-            socket.get().close();
-            socket.set(null);
-            flush();
-            deliverLifecycleAnnouncement(logger, false);
         }
-    }
 
-    /**
-     * Get the set of DJ Link devices which currently can be seen on the network. These can be passed to
-     * {@link VirtualCdj#getLatestStatusFor(DeviceUpdate)} to find the current detailed status for that device,
-     * as long as the Virtual CDJ is active.
-     *
-     * @return the devices which have been heard from recently enough to be considered present on the network
-     *
-     * @throws IllegalStateException if the {@code DeviceFinder} is not active
-     */
-    public Set<DeviceAnnouncement> getCurrentDevices() {
-        if (!isRunning()) {
-            throw new IllegalStateException("DeviceFinder is not active");
-        }
-        expireDevices();  // Get rid of anything past its sell-by date.
-        // Make a copy so callers get an immutable snapshot of the current state.
-        return Collections.unmodifiableSet(new HashSet<DeviceAnnouncement>(devices.values()));
+        expireDevices();
```

Comment:
This isn’t going to happen often enough, because you are only calling this method when a device announcement packet is received. This relates to why I was saying the logic about adjusting the socket timeout in `UpdateSocketConnection` didn’t belong there. We are either going to have to do all the logic of deciding when devices are timing out in that class, or this thread is going to need to have its own thread that periodically expires devices we haven’t heard from recently enough, which means you will need to bring back the `start()` and `stop()` methods (which you have to do anyway, because please do not break backwards compatibility anywhere that we are absolutely unable to avoid it), and this class will need to register as a `LifecycleListener` on `AnnouncementSocketConnection` so that it can shut itself down whenever that class is stopping.

Reply:
To clarify what I mean here, suppose the network cable gets yanked out. No more announcement packets will be received, and we won’t know that we are supposed to expire the devices that we are no longer hearing from because this method will only be called when we get such a packet. So we’ll need a thread that is sleeping for the shorter interval that was being used in the `receive()` call when devices were present in our list, and when it wakes up we will call `expireDevices()`, whether or not any packets have been received.